src/java/org/apache/cassandra/index/sai/disk/v1/postings/PostingsReader.java

```diff
@@ -0,0 +1,395 @@
+            input.close();
+        }
+        finally
+        {
+            summary.close();
+        }
+    }
+
+    @Override
+    public long size()
+    {
+        return numPostings;
+    }
+
+    /**
+     * Advances to the first row ID beyond the current that is greater than or equal to the
+     * target, and returns that row ID. Exhausts the iterator and returns {@link #END_OF_STREAM} if
+     * the target is greater than the highest row ID.
+     *
+     * Does binary search over the skip table to find the next block to load into memory.
+     *
+     * Note: Callers must use the return value of this method before calling {@link #nextPosting()}, as calling
+     * that method will return the next posting, not the one to which we have just advanced.
+     *
+     * @param targetRowID target row ID to advance to
+     *
+     * @return first segment row ID which is >= the target row ID or {@link PostingList#END_OF_STREAM} if one does not exist
+     */
+    @Override
+    public long advance(long targetRowID) throws IOException
+    {
+        listener.onAdvance();
+        int block = binarySearchBlock(targetRowID);
+
+        if (block < 0)
+        {
+            block = -block - 1;
+        }
+
+        if (postingsBlockIdx == block + 1)
+        {
+            // we're in the same block, just iterate through
+            return slowAdvance(targetRowID);
+        }
+        assert block > 0;
+        // Even if there was an exact match, block might contain duplicates.
+        // We iterate to the target token from the beginning.
+        lastPosInBlock(block - 1);
+        return slowAdvance(targetRowID);
+    }
+
+    private long slowAdvance(long targetRowID) throws IOException
+    {
+        while (totalPostingsRead < numPostings)
+        {
+            long segmentRowId = peekNext();
+
+            advanceOnePosition(segmentRowId);
+
+            if (segmentRowId >= targetRowID)
+            {
+                return segmentRowId;
+            }
+        }
+        return END_OF_STREAM;
+    }
+
+    private int binarySearchBlock(long targetRowID)
+    {
+        int low = postingsBlockIdx - 1;
+        int high = Math.toIntExact(blockMaxValues.length()) - 1;
+
+        // in current block
+        if (low <= high && targetRowID <= blockMaxValues.get(low))
+            return low;
+
+        while (low <= high)
+        {
+            int mid = low + ((high - low) >> 1) ;
+
+            long midVal = blockMaxValues.get(mid);
+
+            if (midVal < targetRowID)
+            {
+                low = mid + 1;
+            }
+            else if (midVal > targetRowID)
+            {
+                high = mid - 1;
+            }
+            else
+            {
+                // target found, but we need to check for duplicates
+                if (mid > 0 && blockMaxValues.get(mid - 1L) == targetRowID)
+                {
+                    // there are duplicates, pivot left
+                    high = mid - 1;
```

Comment:
nit: Any tests that hit this path?

Reply:
No. I am in the process of adding testing around this class and the `BlockPackedReader` classes as well as performing a bit of refactoring / renaming to make these classes easier to understand.

Reply:
I have added a test to `PostingsTest` that exercises this condition and I have done a bit of renaming of the fields in this method and have documented it a bit more to make it clearer what is going on here. This is very specific and quite difficult to reproduce. As in, it needs very specific conditions to be triggered but it does need to be there to catch those conditions.